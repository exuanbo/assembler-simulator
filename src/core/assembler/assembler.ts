import { expectNever } from 'ts-expect'

import { ComposeProvider } from '@/common/utils/context'
import { invariant } from '@/common/utils/invariant'

import { AssemblerState, createAssemblerState, useAssemblerState } from './assembler.state'
import { getSize, hasIdentifier, type WithIdentifier } from './assembler.utils'
import {
  type AssemblyNode,
  type AssemblyUnit,
  type CodeChunk,
  finalize,
  initUnit,
} from './assemblyunit'
import * as AST from './ast'
import { AssemblerError, type AssemblyError, ErrorCode, ParserError, Severity } from './errors'
import * as InstrSet from './instrset'
import { resolveOpcode } from './instrset.utils'
import { createLexer } from './lexer'
import { createParser } from './parser'
import { createParserContext, ParserContext, useParserContext } from './parser.context'
import { createTokenStream, TokenStream } from './token.stream'

interface PendingChunk extends CodeChunk {
  node: WithIdentifier<ProcessableNode>
}

type ProcessableNode = Extract<AST.Statement, AssemblyNode>

type ResolvableNode =
  ProcessableNode extends { children: (infer Values)[] }
    ? Exclude<Values, AST.Mnemonic>
    : never

export class Assembler {
  private pendings!: PendingChunk[]
  private unit!: AssemblyUnit

  private setup(): AssemblyUnit {
    this.pendings = []
    return (this.unit = initUnit())
  }

  assemble(input: string): AssemblyUnit {
    const lexer = createLexer(input)
    const stream = createTokenStream(lexer)
    const context = createParserContext()
    const initialState = createAssemblerState()
    return ComposeProvider({
      contexts: [
        AssemblerState.Provider({ value: initialState }),
        ParserContext.Provider({ value: context }),
        TokenStream.Provider({ value: stream }),
      ],
      callback: () => {
        const unit = this.setup()
        const ast = this.parse()
        return this.hasError()
          ? finalize(unit, { chunks: [] })
          : finalize(unit, { ast, labels: context.labels })
      },
    })
  }

  private parse(): AST.Program | null {
    const context = useParserContext()
    const parser = createParser()
    try {
      while (true) {
        const { done, value: node } = parser.next()
        if (done) {
          return node
        }
        this.collectErrors(context)
        this.processStatement(node)
      }
    }
    catch (error) {
      if (error instanceof ParserError) {
        this.addError(error)
      }
      else throw error
    }
    finally {
      context.checkLabels()
      this.collectErrors(context)
      this.processPendings()
    }
    return null
  }

  private hasError(): boolean {
    return !!this.unit.errors.length
  }

  private collectErrors(context: ParserContext): void {
    const errors = context.flushErrors()
    this.addError(...errors)
  }

  private catchError<Result>(callback: (() => Result)): Result | undefined
  private catchError<Result, Fallback>(callback: (() => Result), fallback: Fallback): Result | Fallback
  private catchError<Result, Fallback>(callback: (() => Result), fallback?: Fallback) {
    try {
      return callback()
    }
    catch (error) {
      if (error instanceof AssemblerError) {
        this.addError(error)
      }
      else throw error
    }
    return fallback
  }

  private addError(...errors: AssemblyError[]): void {
    errors.forEach((error) => {
      if (error.severity === Severity.Error) {
        this.unit.errors.push(error)
      }
      else {
        this.unit.warnings.push(error)
      }
    })
  }

  private updateAddress(node: AST.Immediate | ProcessableNode): void {
    const state = useAssemblerState()
    if (node.type === AST.NodeType.Immediate) {
      this.catchError(() => state.setAddress(node))
    }
    else {
      this.catchError(() => state.advanceAddress(node))
    }
  }

  private processStatement(node: AST.Statement): void {
    switch (node.type) {
    case AST.NodeType.Label:
      this.processLabel(node.identifier)
      break
    case AST.NodeType.Instruction:
      this.processInstruction(node)
      this.updateAddress(node)
      break
    case AST.NodeType.Directive:
      switch (node.name) {
      case AST.DirectiveName.DB:
        this.processDataByte(node)
        this.updateAddress(node)
        break
      case AST.DirectiveName.ORG:
        this.updateAddress(node.address)
        break
      }
      break
    }
  }

  private processLabel({ children: [name] }: AST.Identifier): void {
    const state = useAssemblerState()
    const context = useParserContext()
    const label = context.labels.get(name)
    invariant(label?.loc, `Label '${name}' is undefined`)
    label.address = state.address
  }

  private processInstruction(node: AST.Instruction): void {
    const state = useAssemblerState()
    const buffer = new Uint8Array(getSize(node))
    const baseChunk = {
      offset: state.address,
      buffer,
    }
    if (hasIdentifier(node)) {
      const pending = Object.assign(baseChunk, { node })
      this.pendings.push(pending)
      this.unit.chunks.push(pending)
    }
    else {
      const chunk = Object.assign(baseChunk, { node })
      this.encodeInstruction(node, chunk.buffer)
      this.unit.chunks.push(chunk)
    }
  }

  private processDataByte(node: AST.Db): void {
    const state = useAssemblerState()
    const buffer = new Uint8Array(this.resolve(node.children))
    this.unit.chunks.push({
      offset: state.address,
      buffer,
      node,
    })
  }

  private processPendings(): void {
    this.pendings.forEach((chunk) => {
      const state = createAssemblerState(chunk.offset)
      AssemblerState.Provider({
        value: state,
        callback: () => this.encodePending(chunk),
      })
    })
  }

  private encodePending({ buffer, node }: PendingChunk): void {
    switch (node.type) {
    case AST.NodeType.Instruction:
      this.encodeInstruction(node, buffer)
      break
    default:
      expectNever(node.type)
    }
  }

  private encodeInstruction(node: AST.Instruction, buffer: Uint8Array): void {
    const [mnemonic, ...operands] = node.children
    const opcode = resolveOpcode(mnemonic, operands)
    buffer.set([opcode, ...this.resolve(operands)])
  }

  private resolve(nodes: ResolvableNode[]): number[] {
    return this.catchError(() => unsafe_resolve(nodes), [])
  }
}

function unsafe_resolve(nodes: ResolvableNode[]): number[] {
  return nodes.flatMap((node) => {
    switch (node.type) {
    case AST.NodeType.MemoryOperand:
      return unsafe_resolve(node.children)
    case AST.NodeType.Identifier:
      return resolveIdentifier(node)
    case AST.NodeType.Register:
      return resolveRegister(node)
    case AST.NodeType.Immediate:
      return resolveImmediate(node)
    case AST.NodeType.StringLiteral:
      return resolveStringLiteral(node)
    }
  })
}

function resolveIdentifier({ children: [name], loc }: AST.Identifier): number {
  const state = useAssemblerState()
  const context = useParserContext()
  const label = context.labels.get(name)
  invariant(label, `Label '${name}' is not added`)
  invariant(!Number.isNaN(label.address), `Label '${name}' is not processed`)
  const offset = label.address - state.address
  if ((offset < MIN_SIGNED8) || (offset > MAX_SIGNED8)) {
    throw new AssemblerError(ErrorCode.JumpOutOfRange, loc, { offset })
  }
  return offset
}

function resolveRegister({ children: [name] }: AST.Register): number {
  return InstrSet.Register[name]
}

function resolveImmediate({ children: [value], loc }: AST.Immediate): number {
  if (value > MAX_UNSIGNED8) {
    throw new AssemblerError(ErrorCode.ImmediateOutOfRange, loc, { value })
  }
  return value
}

function resolveStringLiteral({ children: chars, loc }: AST.StringLiteral): number[] {
  return Array.from(chars, (char, i) => {
    const value = chars.charCodeAt(i)
    if (value > MAX_UNSIGNED8) {
      throw new AssemblerError(ErrorCode.CharacterOutOfRange, loc, { char, value })
    }
    return value
  })
}

// TODO: move to separate file
const MIN_SIGNED8 =  -0b10000000
const MAX_SIGNED8 =   0b01111111
const MAX_UNSIGNED8 = 0b11111111
