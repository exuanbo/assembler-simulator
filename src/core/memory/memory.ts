import { filter, map, type Observable, share, tap } from 'rxjs'

import { type Bus, Control, type Signals } from '../bus/bus'

export class Memory {
  // TODO: use shared constants
  private readonly data = new Uint8Array(0x100)

  private readonly read$: Observable<Signals>
  private readonly write$: Observable<Signals>

  constructor(
    private readonly bus: Bus,
  ) {
    this.read$ = this.bus.signals$.pipe(
      filter((signals) => signals.control === Control.MEMORY_READ),
      tap(this.read),
      share(),
    )

    this.write$ = this.bus.signals$.pipe(
      filter((signals) => signals.control === Control.MEMORY_WRITE),
      tap(this.write),
      share(),
    )

    this.read$.subscribe()
    this.write$.subscribe()
  }

  private read = (signals: Signals) => {
    this.bus.put({
      data: this.data[signals.address],
      control: Control.CLOCK_IDLE,
    })
  }

  private write = (signals: Signals) => {
    this.data[signals.address] = signals.data
    this.bus.put({
      control: Control.CLOCK_IDLE,
    })
  }

  get data$() {
    return this.write$.pipe(map(() => this.getData()))
  }

  getData() {
    return Array.from(this.data)
  }

  load(data: Uint8Array, offset: number) {
    this.data.set(data, offset)
  }

  reset() {
    this.data.fill(0)
  }
}
