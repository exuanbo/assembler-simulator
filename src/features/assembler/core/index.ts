import { Mnemonic } from '@/common/constants'
import { invariant, unsign8 } from '@/common/utils'

import {
  AssembleEndOfMemoryError,
  DuplicateLabelError,
  JumpDistanceError,
  LabelNotExistError,
} from './exceptions'
import { type Operand, OperandType, parse, type Statement } from './parser'

export type { AssemblerErrorObject } from './exceptions'
export { AssemblerError } from './exceptions'
export type { Statement } from './parser'
export type { SourceRange } from './types'

interface LabelToAddressMap {
  [labelIdentifier: string]: number
}

const getLabelToAddressMap = (statements: Statement[]): LabelToAddressMap => {
  const labelToAddressMap: LabelToAddressMap = {}
  const statementCount = statements.length
  for (let address = 0, statementIndex = 0; statementIndex < statementCount; statementIndex++) {
    const statement = statements[statementIndex]
    const { label, instruction, operands, codes } = statement
    if (label !== null) {
      if (label.identifier in labelToAddressMap) {
        throw new DuplicateLabelError(label)
      }
      labelToAddressMap[label.identifier] = address
    }
    const firstOperand = operands[0] as Operand | undefined
    if (instruction.mnemonic === Mnemonic.ORG) {
      invariant(typeof firstOperand?.code === 'number')
      address = firstOperand.code
    } else {
      // label value has not been calculated yet
      address += codes.length + (firstOperand?.type === OperandType.Label ? 1 : 0)
      if (address > 0xff && statementIndex !== statementCount - 1) {
        throw new AssembleEndOfMemoryError(statement)
      }
    }
  }
  return labelToAddressMap
}

export interface AddressToCodeMap {
  [address: number]: number
}

export interface AddressToStatementMap {
  [address: number]: Statement
}

export type AssembleResult = [AddressToCodeMap, Partial<AddressToStatementMap>]

export const assemble = (source: string): AssembleResult => {
  const statements = parse(source)
  const labelToAddressMap = getLabelToAddressMap(statements)
  const addressToCodeMap: AddressToCodeMap = {}
  const addressToStatementMap: AddressToStatementMap = {}
  const statementCount = statements.length
  for (let address = 0, statementIndex = 0; statementIndex < statementCount; statementIndex++) {
    const statement = statements[statementIndex]
    const { instruction, operands, codes } = statement
    const firstOperand = operands[0] as Operand | undefined
    if (instruction.mnemonic === Mnemonic.ORG) {
      invariant(typeof firstOperand?.code === 'number')
      address = firstOperand.code
      continue
    }
    if (firstOperand?.type === OperandType.Label) {
      if (!(firstOperand.value in labelToAddressMap)) {
        throw new LabelNotExistError(firstOperand)
      }
      const distance = labelToAddressMap[firstOperand.value] - address
      if (distance < -128 || distance > 127) {
        throw new JumpDistanceError(firstOperand)
      }
      const unsignedDistance = unsign8(distance)
      firstOperand.code = unsignedDistance
      codes.push(unsignedDistance)
    }
    const nextAddress = address + codes.length
    codes.forEach((code, codeIndex) => {
      addressToCodeMap[address + codeIndex] = code
    })
    addressToStatementMap[address] = statement
    address = nextAddress
  }
  return [addressToCodeMap, addressToStatementMap]
}
