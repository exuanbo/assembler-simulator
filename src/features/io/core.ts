export const SKIP = -1

export const MAX_PORT = 0x0f

export enum InputPort {
  SimulatedKeyboard = 0,
  Thermostat = 3,
  Keyboard = 7,
  NumericKeypad = 8,
}

export type InputData =
  | {
      content: number
      port: InputPort
    }
  | {
      content: null
      port: null
    }

export interface InputSignals {
  data: InputData
  interrupt: boolean
  // TODO: interruptVectorAddress
}

export const initialInputSignals: InputSignals = {
  data: {
    content: null,
    port: null,
  },
  interrupt: false,
}

export enum OutputPort {
  TrafficLights = 1,
  SevenSegmentDisplay = 2,
  Heater = 3,
  SnakeInMaze = 4,
  StepperMotor = 5,
  Lift = 6,
  Keyboard = 7,
  NumericKeypad = 8,
}

type OutputFlagSignalName = 'halted' | 'closeWindows'

type OutputFlagSignals = {
  [signalName in OutputFlagSignalName]?: true
}

interface OutputData {
  content: number
  port: OutputPort
}

export interface OutputSignals extends OutputFlagSignals {
  expectedInputPort?: InputPort
  data?: OutputData
}

export interface Signals {
  input: InputSignals
  output: OutputSignals
}
