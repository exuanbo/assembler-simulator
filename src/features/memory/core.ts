import { Ascii, Mnemonic } from '@/common/constants'
import type { AddressToCodeMap, AddressToStatementMap } from '@/features/assembler/core'

export type MemoryData = number[]

const MEMORY_SIZE = 0x100

export const VDU_START_ADDRESS = 0xc0

export const initVduData = (): number[] =>
  new Array<number>(MEMORY_SIZE - VDU_START_ADDRESS).fill(Ascii.Space)

export const initData = (): MemoryData =>
  new Array<number>(VDU_START_ADDRESS).fill(0).concat(initVduData())

export const initDataFrom = (map: AddressToCodeMap): MemoryData => {
  const data = initData()
  for (const address in map) {
    data[address] = map[address]
  }
  return data
}

export const getVduDataFrom = (data: MemoryData): number[] => data.slice(VDU_START_ADDRESS)

export const getSourceFrom = (map: Partial<AddressToStatementMap>): string[] => {
  const source: string[] = []
  for (let address = 0; address < MEMORY_SIZE; address++) {
    source.push(address < VDU_START_ADDRESS ? Mnemonic.END : '')
  }
  for (const address in map) {
    const statement = map[address]
    // istanbul ignore next
    if (statement === undefined) {
      continue
    }
    const { instruction, operands } = statement
    if (instruction.mnemonic === Mnemonic.DB) {
      const operand = operands[0]
      // OperandType.Number
      if (typeof operand.code === 'number') {
        source[address] = operand.value
      } else {
        // OperandType.String
        operand.value.split('').forEach((char, charIndex) => {
          source[Number(address) + charIndex] = char
        })
      }
    } else {
      source[address] = instruction.mnemonic
      const nextAddress = Number(address) + 1
      operands.forEach((operand, operandIndex) => {
        const { value } = operand
        // Address or RegisterAddress
        const isAddressOperand = operand.type.endsWith('Address')
        source[nextAddress + operandIndex] = isAddressOperand ? `[${value}]` : value
      })
    }
  }
  return source
}
