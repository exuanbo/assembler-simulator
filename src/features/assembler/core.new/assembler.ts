import { ComposeProvider } from '@/common/utils/context'
import { invariant } from '@/common/utils/invariant'

import { AssemblerState, createAssemblerState, useAssemblerState } from './assembler.state'
import { hasIdentifier, type WithIdentifier } from './assembler.utils'
import {
  type AssemblyNode,
  type AssemblyUnit,
  type CodeChunk,
  finish,
  initUnit,
} from './assemblyunit'
import * as AST from './ast'
import { AssemblerError, type AssemblyError, ErrorCode, ParserError, Severity } from './errors'
import * as InstrSet from './instrset'
import { resolveOpcode } from './instrset.utils'
import { createLexer } from './lexer'
import { createParser } from './parser'
import { createParserContext, ParserContext, useParserContext } from './parser.context'
import { createTokenStream, TokenStream } from './token.stream'

// TODO: move to somewhere
const MAX_NUMBER = 0xff

export interface Assembler {
  run(input: string): AssemblyUnit
}

export function createAssembler(): Assembler {
  return new AssemblerImpl()
}

interface PendingChunk extends CodeChunk {
  node: WithIdentifier<ProcessableNode>
}

type ProcessableNode = Extract<AST.Statement, AssemblyNode>

type ResolvableNode =
  ProcessableNode extends { children: (infer Values)[] }
    ? Exclude<Values, AST.Mnemonic>
    : never

class AssemblerImpl implements Assembler {
  private pending!: PendingChunk[]
  private unit!: AssemblyUnit

  private setup(): AssemblyUnit {
    this.pending = []
    return (this.unit = initUnit())
  }

  run(input: string): AssemblyUnit {
    const lexer = createLexer(input)
    const stream = createTokenStream(lexer)
    const context = createParserContext()
    const initialState = createAssemblerState()
    return ComposeProvider({
      contexts: [
        AssemblerState.Provider({ value: initialState }),
        ParserContext.Provider({ value: context }),
        TokenStream.Provider({ value: stream }),
      ],
      callback: () => {
        const unit = this.setup()
        const ast = this.parse()
        return this.hasError()
          ? finish(unit, { chunks: [] })
          : finish(unit, { ast, labels: context.labels })
      },
    })
  }

  private parse(): AST.Program | null {
    const context = useParserContext()
    const parser = createParser()
    try {
      while (true) {
        const { done, value: node } = parser.next()
        if (done) {
          return node
        }
        this.collectErrors(context)
        this.processStatement(node)
      }
    }
    catch (error) {
      if (error instanceof ParserError) {
        this.addError(error)
      }
      else throw error
    }
    finally {
      context.checkLabels()
      this.collectErrors(context)
      this.processPendings()
    }
    return null
  }

  private hasError(): boolean {
    return !!this.unit.errors.length
  }

  private collectErrors(context: ParserContext): void {
    const errors = context.flushErrors()
    this.addError(...errors)
  }

  private catchError<Result>(callback: (() => Result)): Result | undefined
  private catchError<Result, Fallback>(callback: (() => Result), fallback: Fallback): Result | Fallback
  private catchError<Result, Fallback>(callback: (() => Result), fallback?: Fallback) {
    try {
      return callback()
    }
    catch (error) {
      if (error instanceof AssemblerError) {
        this.addError(error)
      }
      else throw error
    }
    return fallback
  }

  private addError(...errors: AssemblyError[]): void {
    errors.forEach((error) => {
      if (error.severity === Severity.Error) {
        this.unit.errors.push(error)
      }
      else {
        this.unit.warnings.push(error)
      }
    })
  }

  private updateAddress(node: AST.Immediate | ProcessableNode): void {
    const state = useAssemblerState()
    if (node.type === AST.NodeType.Immediate) {
      this.catchError(() => state.setAddress(node))
    }
    else {
      this.catchError(() => state.advanceAddress(node))
    }
  }

  private processStatement(node: AST.Statement): void {
    switch (node.type) {
    case AST.NodeType.Label:
      this.processLabel(node.identifier)
      break
    case AST.NodeType.Instruction:
      this.processInstruction(node)
      this.updateAddress(node)
      break
    case AST.NodeType.Directive:
      switch (node.name) {
      case AST.DirectiveName.DB:
        this.processDataByte(node)
        this.updateAddress(node)
        break
      case AST.DirectiveName.ORG:
        this.updateAddress(node.address)
        break
      }
      break
    }
  }

  private processLabel({ children: [id] }: AST.Identifier): void {
    const state = useAssemblerState()
    const context = useParserContext()
    const label = context.labels.get(id)
    invariant(label?.loc, `Label '${id}' is undefined`)
    label.address = state.address
  }

  private processInstruction(node: AST.Instruction): void {
    const state = useAssemblerState()
    const chunk = {
      address: state.address,
      codes: [],
    }
    if (hasIdentifier(node)) {
      this.pending.push({ ...chunk, node })
    }
    else {
      this.encodeInstruction(node, chunk.codes)
      this.unit.chunks.push({ ...chunk, node })
    }
  }

  private processDataByte(node: AST.Db): void {
    const state = useAssemblerState()
    const codes = this.resolve(node.children)
    this.unit.chunks.push({
      address: state.address,
      codes,
      node,
    })
  }

  private processPendings(): void {
    this.pending.forEach((chunk) => {
      AssemblerState.Provider({
        value: createAssemblerState(chunk.address),
        callback: () => this.encodePending(chunk),
      })
      this.unit.chunks.push(chunk)
    })
  }

  private encodePending({ codes, node }: PendingChunk): void {
    switch (node.type) {
    case AST.NodeType.Instruction:
      this.encodeInstruction(node, codes)
      break
    default:
      // @ts-expect-error: never
      throw new Error(`Node '${node.type.toString()}' is not processed`)
    }
  }

  private encodeInstruction(node: AST.Instruction, codes: number[]): void {
    const [mnemonic, ...operands] = node.children
    const opcode = resolveOpcode(mnemonic, operands)
    codes.push(opcode, ...this.resolve(operands))
  }

  private resolve(nodes: ResolvableNode[]): number[] {
    return this.catchError(() => unsafe_resolve(nodes), [])
  }
}

function unsafe_resolve(nodes: ResolvableNode[]): number[] {
  return nodes.flatMap((node) => {
    switch (node.type) {
    case AST.NodeType.MemoryOperand:
      return unsafe_resolve(node.children)
    case AST.NodeType.Identifier:
      return resolveIdentifier(node)
    case AST.NodeType.Register:
      return resolveRegister(node)
    case AST.NodeType.Immediate:
      return resolveImmediate(node)
    case AST.NodeType.StringLiteral:
      return resolveStringLiteral(node)
    }
  })
}

function resolveIdentifier({ children: [id], loc }: AST.Identifier): number {
  const state = useAssemblerState()
  const context = useParserContext()
  const label = context.labels.get(id)
  invariant(label, `Label '${id}' is not added`)
  invariant(!Number.isNaN(label.address), `Label '${id}' is not processed`)
  const distance = label.address - state.address
  if ((distance < -0x80) || (distance > 0x7f)) {
    throw new AssemblerError(ErrorCode.JumpOutOfRange, loc, { distance })
  }
  return unsign8(distance)
}

// TODO: move to utils
function unsign8(value: number): number {
  return (value < 0) ? (value + MAX_NUMBER + 1) : value
}

function resolveRegister({ children: [name] }: AST.Register): number {
  return InstrSet.Register[name]
}

function resolveImmediate({ children: [value], loc }: AST.Immediate): number {
  if (value > MAX_NUMBER) {
    throw new AssemblerError(ErrorCode.ImmediateOutOfRange, loc, { value })
  }
  return value
}

function resolveStringLiteral({ children: chars, loc }: AST.StringLiteral): number[] {
  return Array.from(chars, (char, i) => {
    const value = chars.charCodeAt(i)
    if (value > MAX_NUMBER) {
      throw new AssemblerError(ErrorCode.CharacterOutOfRange, loc, { char, value })
    }
    return value
  })
}
