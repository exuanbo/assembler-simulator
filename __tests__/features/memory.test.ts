import { assemble } from '@/features/assembler/core'
import { getSourceFrom, getVduDataFrom, initData } from '@/features/memory/core'

import { memoryDataSerializer } from '../snapshotSerializers'

expect.addSnapshotSerializer(memoryDataSerializer)

describe('memory', () => {
  describe('getVduDataFrom', () => {
    it('should return the VDU data', () => {
      const memoryData = initData()
      const vduData = getVduDataFrom(memoryData)
      expect(vduData).toMatchSnapshot()
    })
  })

  describe('getSourceFrom', () => {
    it('should return correct source', () => {
      const [, addressToStatementMap] = assemble('jmp start start: add al, 01 mov bl, [cl] end')
      const source = getSourceFrom(addressToStatementMap)
      expect(source).toMatchSnapshot()
    })

    describe('with DB', () => {
      it('should display number', () => {
        const [, addressToStatementMap] = assemble('db a0 end')
        const source = getSourceFrom(addressToStatementMap)
        expect(source).toMatchSnapshot()
      })

      it('should split string', () => {
        const [, addressToStatementMap] = assemble('db "Hello, world!" end')
        const source = getSourceFrom(addressToStatementMap)
        expect(source).toMatchSnapshot()
      })
    })
  })
})
