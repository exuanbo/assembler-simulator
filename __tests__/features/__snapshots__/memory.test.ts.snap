// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`memory getSourceFrom should return correct source 1`] = `
[
  JMP, START, ADD, AL, 01, MOV, BL, [CL], END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  , , , , , , , , , , , , , , , ,
  , , , , , , , , , , , , , , , ,
  , , , , , , , , , , , , , , , ,
  , , , , , , , , , , , , , , , ,
]
`;

exports[`memory getSourceFrom with DB should display number 1`] = `
[
  A0, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  , , , , , , , , , , , , , , , ,
  , , , , , , , , , , , , , , , ,
  , , , , , , , , , , , , , , , ,
  , , , , , , , , , , , , , , , ,
]
`;

exports[`memory getSourceFrom with DB should split string 1`] = `
[
  H, e, l, l, o, ,,  , w, o, r, l, d, !, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  END, END, END, END, END, END, END, END, END, END, END, END, END, END, END, END,
  , , , , , , , , , , , , , , , ,
  , , , , , , , , , , , , , , , ,
  , , , , , , , , , , , , , , , ,
  , , , , , , , , , , , , , , , ,
]
`;

exports[`memory getVduDataFrom should return the VDU data 1`] = `
[
  20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20,
  20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20,
  20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20,
  20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20, 20,
]
`;
